src/python/pants/backend/jvm/tasks/analysis_extraction.py

```diff
@@ -47,106 +49,122 @@ def _create_products_if_should_run(self):
+      '1.20.0.dev2',
+      'The `classes_by_source` product depends on internal compiler details and is no longer produced.'
+    )
+
     if not self._create_products_if_should_run():
       return
 
-    zinc_analysis = self.context.products.get_data('zinc_analysis')
     classpath_product = self.context.products.get_data('runtime_classpath')
-    classes_by_source = self.context.products.get_data('classes_by_source')
-    product_deps_by_src = self.context.products.get_data('product_deps_by_src')
+    product_deps_by_target = self.context.products.get_data('product_deps_by_target')
 
     fingerprint_strategy = DependencyContext.global_instance().create_fingerprint_strategy(
         classpath_product)
 
-    targets = list(zinc_analysis.keys())
+    # classpath fingerprint strategy only works on targets with a classpath.
+    targets = [target for target in self.context.targets() if hasattr(target, 'strict_deps')]
+    potential_deps_classpaths = [
+      entry for target in targets for _, entry in classpath_product.get_for_target(target) if entry
+    ]
     with self.invalidated(targets,
                           fingerprint_strategy=fingerprint_strategy,
                           invalidate_dependents=True) as invalidation_check:
-      # Extract and parse products for any relevant targets.
       for vt in invalidation_check.all_vts:
-        summary_json_file = self._summary_json_file(vt)
-        cp_entry, _, analysis_file = zinc_analysis[vt.target]
+        # class paths for the artifacts created by the target we are computing deps for
+        target_artifact_classpaths = [entry for _, entry in classpath_product.get_for_target(vt.target)]
+
+        jdeps_output_json = self._jdeps_output_json(vt)
         if not vt.valid:
-          self._extract_analysis(vt.target, analysis_file, summary_json_file)
+          self._run_jdeps_analysis(vt.target, target_artifact_classpaths, potential_deps_classpaths, jdeps_output_json)
         self._register_products(vt.target,
-                                cp_entry,
-                                summary_json_file,
-                                classes_by_source,
-                                product_deps_by_src)
-
-  def _extract_analysis(self, target, analysis_file, summary_json_file):
-    target_classpath = self._zinc.compile_classpath('runtime_classpath', target)
-    analysis_by_cp_entry = self._analysis_by_runtime_entry
-    upstream_analysis = list(self._upstream_analysis(target_classpath, analysis_by_cp_entry))
+                                jdeps_output_json,
+                                product_deps_by_target)
+
+  @memoized_property
+  def _jdeps_summary_line_regex(self):
+    return re.compile(r"^.+\s->\s(.+)$")
+
+  def _run_jdeps_analysis(self, target, target_artifact_classpaths, potential_deps_classpaths, jdeps_output_json):
+    with self.aliased_classpaths(potential_deps_classpaths) as classpaths_by_alias:
```

Comment:
So, I think rather than passing in this entire collection, we should only be aliasing the valid classpath entries for the target.

In this case, that would look like using the Zinc subsystem to compute the deps per-target we are going to invoke jdeps on: https://github.com/pantsbuild/pants/blob/master/src/python/pants/backend/jvm/subsystems/zinc.py#L436

Reply:
I understand the false positive issue, but isn't the point of this task to remove reliance on zinc? Maybe we could sort the classpath entries so the ones that come from explicit deps of the target, and their transients, are resolved first? One benefit of using all the classpaths is you can find out if some deps are being used implicitly.

Reply:
The goal is to remove reliance on usage of the zinc tool: but this classpath-computation codepath is our code, and will likely just be renamed at some point.

Reply:
Sorry, I must be missing something. If we compute the deps directly from the dependency context, what new information do we learn by apply jdeps to a target? It seems like the only thing we can figure out about a target is whether it has missing or extra deps. Maybe I just answered my own question but I had assumed a big part of what we wanted was to resolve *what* those extra undeclared deps were, which we can't do unless they are on the classpath.

Reply:
The dependency context computes deps based on what is declared in BUILD files (ie, declared target dependencies). `jdeps` computes actual usage of deps (at the classlevel). It's entirely possible to declare a dependency in your BUILD file without using it, and that's exactly what the `./pants dep-usage` task is supposed to help you fix.

Reply:
And with regard to undeclared deps: those are only possible in the case of `strict_deps: False`, and it refers to things that _are_ in your transitive dependencies (ie, they were passed to the compiler), but which were not listed in your declared dependencies (ie, the BUILD file).

Reply:
I preemptively made this change because I figured there was a good explanation. So when CI is green again we should be good to ship. Also, noted that the main purpose of dep-usage in *un*used deps. The jvm-dep-check linting task on the other hand does compare the actual deps with declared deps. Based on the strict_dependencies docs it sounds like transient strict deps that are "exported" are captured in the dependent targets list of the root. If that isn't true then the dep check could fail in some places. It would manifest as a classpath called `not found` being unresolvable to a target.